rfc/0052-sdk3-full-text-search.md

```diff
@@ -566,7 +566,250 @@ A query that matches nothing.
+      "k": 3,
+      "//": "This is a more important consideration than the other vector query, so give it a higher score boost.",
+      "boost": 0.7
+    },
+    {
+      "field": "cityVector",
+      "//": "This is another query such as 'city with good museums'.",
+      "vector": [
+        -0.005610323283,
+        0.023427998167,
+        0.0132511895034,
+        0.03466271757
+      ],
+      "k": 2,
+      "boost": 0.3
+    }
+  ],
+  "//": "The `knn_operator` field is new and determines how multiple vector searches are combined.",
+  "knn_operator": "or",
+  "//": "The `fields`, `sort` and `limit` fields come from `SearchOptions`, as usual.",
+  "fields": [
+    "citName",
+    "cityDesc"
+  ],
+  "sort": [
+    "-_score"
+  ],
+  "limit": 5
+}
+```
+
+Important SDK considerations:
+
+* There is some uncertainty on whether it is mandatory to provide a `query` field containing a 'normal' FTS query.  This is being followed up with the FTS team.  As it makes sense to be able to perform solely a vector search, we will assume for now that will be the case.
+* The `knn` field can contain multiple vector queries.  It must always contain at least 1.
+  * How they are combined is handled with a top-level `knn_operator` parameter.
+* There may be other vector search types in future, such as Approximate Nearest Neighbour (ANN).  So we may want to aim for a level of abstraction and avoid explicitly mentioning `knn` and perhaps `k`.
+* Inside an individual vector query:
+  * It contains a single vector.  FAISS allows multiple vectors to be sent in one query and the FTS team say that may want to be exposed later.
+  * The `field` parameter is essentially mandatory.  While FTS will not reject a query that does not contain it, it prevents that vector search from having any effect. 
+  * The `k` parameter is mandatory.  The FTS team suggest 3 is a sensible default to send from the SDKs.
+* There are new FTS index definition fields for defining the vector index, but they are inside the `plan` top-level field.  
+So there is nothing extra required from the SDKs to support these, as we already simply turn the untyped `plan` parameters into JSON.
+* The SDK (and Couchbase in general), at least in this first phase, have nothing to do with generating the vectors - both those stored in the documents, and those used for queries.
+The user will be responsible for generating these, for example using the OpenAI Embeddings API.
+* The feature works with both global and scoped FTS indexes.
+
+SDK PM advises that the most common vector search use-cases are likely to be, in this order:
+
+* Performing a vector search in isolation (without a traditional FTS query).
+* A SQL++ statement is used to also perform a vector search.
+  * This is just passed in the string statement and is expected to return the same query results JSON as at present, so from the SDK POV it's just a passthrough with no SDK modifications needed.
```

Comment:
@jon-strabala can you confirm that the query results will be the same in this mode?